add synchronization to SupervisorManager

dclim opened the pull request.

gianm commented:
`getSupervisorStatus` doesn't need synchronization, but it does have a race because it does `supervisors.get(id)` twice. It should only do it once.

gianm commented:
@dclim LGTM other than prior comment

dclim commented:
@gianm ah I missed that, thanks

gianm commented:
👍

fjy commented on indexing-service/src/main/java/io/druid/indexing/overlord/supervisor/SupervisorManager.java:
```diff
@@ -56,38 +56,29 @@ public SupervisorManager(MetadataSupervisorManager metadataSupervisorManager)
 
   @LifecycleStart
-  public void start()
+  public synchronized void start()
```
why not have a lock instead of syncing on the entire start method?

dclim replied:
@fjy no particular reason for using the instance's own intrinsic lock and synchronizing whole methods other than minimizing lines of code added; this class isn't really used in any highly concurrent fashion so I wasn't too concerned, but having said that I can definitely use a lock object.

Out of curiosity, how do people in the Druid community generally feel about synchronizing on 'this'? I know some devs have differing opinions on it (i.e. your lock is public and others can steal your lock, etc.)

himanshug replied:
I would be ok with using synchronized(this) in this case, can't see any advantage of using a new lock object. In general, no code should be using instance of SupervisorManager etc as a lock and using "this" should be ok then.

fjy commented on indexing-service/src/main/java/io/druid/indexing/overlord/supervisor/SupervisorManager.java:
```diff
@@ -101,7 +92,7 @@ public void start()
 
   @LifecycleStop
-  public void stop()
+  public synchronized void stop()
```
same here

fjy commented on indexing-service/src/main/java/io/druid/indexing/overlord/supervisor/SupervisorManager.java:
```diff
@@ -56,38 +56,29 @@ public SupervisorManager(MetadataSupervisorManager metadataSupervisorManager)
   }
 
-  public void stopAndRemoveSupervisor(String id, boolean writeTombstone)
+  public synchronized boolean stopAndRemoveSupervisor(String id)
```
do we need to synchronize on 'this' or can we have a lock?

drcrallen commented on indexing-service/src/main/java/io/druid/indexing/overlord/supervisor/SupervisorManager.java:
```diff
@@ -56,33 +57,27 @@ public SupervisorManager(MetadataSupervisorManager metadataSupervisorManager)
-           ? Optional.<SupervisorSpec>absent()
-           : Optional.fromNullable(supervisors.get(id).rhs);
+    Pair<Supervisor, SupervisorSpec> supervisor = supervisors.get(id);
+    return supervisor == null ? Optional.<SupervisorSpec>absent() : Optional.fromNullable(supervisor.rhs);
```
This method is public. Needs guard against NPE on non existing id.

dclim replied:
@drcrallen doesn't `supervisors.get(non-existing-id)` return null?

drcrallen replied:
Yes, sorry, I was on mobile when I made the comment. You are correct

drcrallen commented on indexing-service/src/main/java/io/druid/indexing/overlord/supervisor/SupervisorManager.java:
```diff
@@ -118,9 +117,24 @@ public void stop()
-           ? Optional.<SupervisorReport>absent()
-           : Optional.fromNullable(supervisors.get(id).lhs.getStatus());
+    Pair<Supervisor, SupervisorSpec> supervisor = supervisors.get(id);
+    return supervisor == null ? Optional.<SupervisorReport>absent() : Optional.fromNullable(supervisor.lhs.getStatus());
```
Also guard against missing id.

dclim replied:
@drcrallen doesn't supervisors.get(non-existing-id) return null?

drcrallen replied:
Yes, is fine

drcrallen commented on indexing-service/src/main/java/io/druid/indexing/overlord/supervisor/SupervisorManager.java:
```diff
@@ -118,9 +117,24 @@ public void stop()
+    return supervisor == null ? Optional.<SupervisorReport>absent() : Optional.fromNullable(supervisor.lhs.getStatus());
+  }
+
+  private boolean possiblyStopAndRemoveSupervisorInternal(String id, boolean writeTombstone)
```
Suggest java doc on expected lock state before calling

dclim replied:
sure can add

drcrallen commented on indexing-service/src/main/java/io/druid/indexing/overlord/supervisor/SupervisorManager.java:
```diff
@@ -118,9 +117,24 @@ public void stop()
+        metadataSupervisorManager.insert(id, new NoopSupervisorSpec()); // where NoopSupervisorSpec is a tombstone
+      }
+      pair.lhs.stop(true);
+      supervisors.remove(id);
```
I think you can move this to the top instead of the get.

dclim replied:
@drcrallen hm, my thought in ordering it this way is that if I removed it from the list first and then tried to stop it, if stop() threw an exception that prevented the supervisor from actually stopping then we would have already removed it from our list and could no longer message it (or we could create a 2nd supervisor that would conflict with the zombie one). With this ordering, it will first remove it from metadata storage (so that even if stop does fail it won't restart on next run), then it'll try to stop it, and only if stop succeeds will it remove it from the list.

For the Kafka implementation, this is mostly moot since stop() shouldn't throw an exception, but since SupervisorManager can manage other types of supervisors as well, it seemed better to make sure it actually stopped before removing it (in case some supervisors could have transient issues that would prevent them from stopping but calling it a 2nd time might succeed?)

drcrallen replied:
Are you expecting stop to throw an exception? the loop in `stop()` for the manager doesn't account for that.

drcrallen replied:
`io.druid.indexing.overlord.supervisor.Supervisor#stop` doesn't explain expected behavior on stop failure, can you coordinate with @gianm to clarify the behavior of the supervisor `stop` method under the kinds of conditions you mentioned?

dclim replied:
@drcrallen that's a good point, I'll change SupervisorManager stop to continue cleaning up supervisors if one of them throws an exception. In general I think implementors should be free to throw a RTE if appropriate if some weirdness happens when they call stop.

drcrallen commented:
@gianm / @dclim this is susceptible to the same thing that bit lookups. Where one bad stop or start can tie up the lock, and then your jetty threads exhaust from stuff waiting on the lock. That should be resolved before merging.

The solution for lookups (at a minimum) is to add qos for thebresource.

Other solutions are to make the management lockless.

drcrallen commented:
There aren't synchronization ordering guarantees when multiple things hold the lock. Will anything bad happen if lots of things are trying to operate on the state, and the locks return in an non-deterministic order?

drcrallen commented on indexing-service/src/main/java/io/druid/indexing/overlord/supervisor/SupervisorManager.java:
```diff
@@ -118,9 +117,24 @@ public void stop()
+  private boolean possiblyStopAndRemoveSupervisorInternal(String id, boolean writeTombstone)
+  {
+    Pair<Supervisor, SupervisorSpec> pair = supervisors.get(id);
+    if (pair != null) {
```
Minor readability change: can you just make this

```java
if(pair == null) {
  return false;
}
```
?

dclim replied:
sure

gianm commented:
@drcrallen the overlord has lots of issues like the one you mentioned (a great many calls involve acquiring coarse locks, so there are a lot of opportunities to lock it up). Since the overlord doesn't do anything other than handle coordination (it doesn't accept queries) I don't think that's a show stopper problem. So I don't think we need to fix that before merging the PR. IMO the lack of thread safety on SupervisorManager is a bigger issue than coarse grained locking.

drcrallen commented:
@gianm can you file a ticket about that and link it here? I'm ok with it being an acknowledged shortcoming since overlord locking up can be very quickly solved with a restart.

dclim commented:
@drcrallen I added exception handling to stop() and added a started flag to prevent the issue you described from occurring. Let me know if you have any other concerns. Thanks for the review!

drcrallen commented:
@dclim Thanks!


Re: https://github.com/druid-io/druid/pull/3077#issuecomment-224009113 if the instance is stopping / stopped while something is trying to add new management stuff, it is possible to have stuff added after stop() has been called. Can you add in a state check to make sure an update can't occur when the manager is in the stopped state?

drcrallen commented on indexing-service/src/main/java/io/druid/indexing/overlord/supervisor/SupervisorManager.java:
```diff
@@ -56,33 +59,29 @@ public SupervisorManager(MetadataSupervisorManager metadataSupervisorManager)
-  public boolean createAndStartSupervisor(SupervisorSpec spec)
+  public boolean createOrUpdateAndStartSupervisor(SupervisorSpec spec)
   {
+    Preconditions.checkState(started, "SupervisorManager not started");
```
You'll want this inside the lock

dclim replied:
@drcrallen if it's inside the lock, won't you have that problem you were talking about where the lifecycle start blocks and Jetty exhausts its threads because this method can't obtain the lock? The reason it's outside of the lock is to address that comment.

drcrallen replied:
There is a race condition between checking and doing something here. As such, the state should *at least* be checked inside the lock. It can fail early as well.

drcrallen replied:
Otherwise, it can be in the process of stopping (while holding the lock) but not assign started = false until after it already thinks it closed everything and after you think it hasn't

dclim replied:
@drcrallen yeah I was aware of that but thought it was such an edge case that it would never happen in practice. If you think about how this code is going to be called, it's a very low concurrency section of code. I can make the stopping case safe too but in my opinion it is probably overkill.

drcrallen replied:
if stop is taking a long time, it will hold the lock for a while and never set started = false. It seems reasonable to me to expect some request to come in and wait on that lock between the time that stop has acquired the lock (and began looping through its stops) and the time it sets started = false

dclim replied:
@drcrallen a fair point; I added another check within the synchronized block. Does that address your concern?

drcrallen replied:
yep thanks

drcrallen commented on indexing-service/src/main/java/io/druid/indexing/overlord/supervisor/SupervisorManager.java:
```diff
@@ -56,33 +59,29 @@ public SupervisorManager(MetadataSupervisorManager metadataSupervisorManager)
-      }
-      pair.lhs.stop(true);
-      supervisors.remove(id);
+    Preconditions.checkState(started, "SupervisorManager not started");
```
Same here, should be inside the lock

drcrallen commented:
Cool, now the fun question: Are you adding unit tests?

dclim commented:
@drcrallen sure I can add some

dclim commented:
@drcrallen added a bunch of tests, thanks for the review!

drcrallen commented:
All this work to add thread-safe methods and no concurrency testing?

I'm still 👍  with the improvements as is though since this is blocking 0.9.1. But please add some more concurrency tests for master soon-ish
